cmd/proxy-to-grafana/proxy-to-grafana.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// proxy-to-grafana uses a reverse proxy which identifies users based on their
+// originating Tailscale node and creates / logs in users to grafana. Uses the
+// auth proxy feature for Grafana:
+// https://grafana.com/docs/grafana/latest/auth/auth-proxy/
+//
+// Provide the env var TS_AUTHKEY to have this server automatically join your
+// tailnet, or look for the logged auth link on first start.
+//
+// Use this Grafana configuration to enable the auth proxy:
+//
+// ```
+// [auth.proxy]
+// enabled = true
+// header_name = X-WEBAUTH-USER
+// header_property = username
+// auto_sign_up = true
+// whitelist = 127.0.0.1
+// headers = Name:X-WEBAUTH-NAME
+// enable_login_token = true
+// ```
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	"inet.af/netaddr"
+	"tailscale.com/client/tailscale"
+	"tailscale.com/net/tsaddr"
+	"tailscale.com/tailcfg"
+	"tailscale.com/tsnet"
+)
+
+var (
+	hostname     = flag.String("hostname", "", "Tailscale hostname to serve on")
+	backendAddr  = flag.String("backend-addr", "", "Address of the Grafana server, in host:port format")
+	tailscaleDir = flag.String("state-dir", "./", "Alternate directory to use for Tailscale state storage. If empty, a default is used.")
+)
+
+func main() {
+	flag.Parse()
+	if *hostname == "" || strings.Contains(*hostname, ".") {
+		log.Fatal("missing or invalid --hostname")
+	}
+	if *backendAddr == "" {
+		log.Fatal("missing --backend-addr")
+	}
+	ts := &tsnet.Server{
+		Dir:      *tailscaleDir,
+		Hostname: *hostname,
+	}
+
+	url, err := url.Parse(fmt.Sprintf("http://%s", *backendAddr))
+	if err != nil {
+		panic(err)
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	originalDirector := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		originalDirector(req)
+		modifyRequest(req)
+	}
+
+	ltsn, err := ts.Listen("tcp", ":80")
```

Comment:
common variable name for listeners is `ln`

Reply:
👍